### PR TITLE
Allow apps to signal whenever there are insufficient scopes

### PR DESCRIPTION
**Pull request by rezaansyed**
Before submitting the PR, please consider if any of the following are needed:

- [ ] Update `CHANGELOG.md` if the changes would impact users
- [ ] Update `README.md`, if appropriate.
- [ ] Update any relevant pages in `docs/`, if necessary
- [ ] For security fixes, the [Disclosure Policy](https://github.com/Shopify/shopify_app/blob/master/SECURITY.md#disclosure-policy) must be followed.

### CODE REVIEW

**Review by theundeadmonk — APPROVED** [review]

File: test/shopify_app/controller_concerns/login_protection_test.rb [hunk 1]
```diff
@@ -462,6 +468,54 @@ class LoginProtectionControllerTest < ActionController::TestCase
...
+      token: token,
+      api_version: '2020-01',
+      extra: { scopes: ShopifyApp.configuration.scope.split(',') }
+    )
+
+    ShopifyApp::SessionRepository.expects(:retrieve_shop_session_by_shopify_domain)
+      .with(domain).returns(expected_session)
+
+    with_application_test_routes do
+      request.env['jwt.shopify_domain'] = domain
```
Comment (paulomarg) [inline test/shopify_app/controller_concerns/login_protection_test.rb:511]: Would it make sense to test that this happens both with and without JWT?

**Review by paulomarg — APPROVED** [review]
LGTM!

File: app/controllers/shopify_app/callback_controller.rb [hunk 2]
```diff
@@ -78,6 +78,10 @@ def token
       auth_hash['credentials']['token']
     end
 
+    def shop_scopes
+      auth_hash['extra']['scope'].split(',')
```
Comment (andyw8) [inline app/controllers/shopify_app/callback_controller.rb:82]: Could `auth_hash['extra']['scope']` ever be nil here?

Comment (rezaansyed) [reply app/controllers/shopify_app/callback_controller.rb:82]: Only if there are no scopes associated with a shopify_session (shop/user). I guess we can add a check for that as well

File: app/controllers/shopify_app/callback_controller.rb [hunk 3]
```diff
@@ -93,9 +97,13 @@ def set_shopify_session
       session[:shopify_user] = associated_user
       if session[:shopify_user].present?
         session[:shop_id] = nil if shop_session && shop_session.domain != shop_name
-        session[:user_id] = ShopifyApp::SessionRepository.store_user_session(session_store, associated_user)
+
+        ActiveRecord::Base.transaction do
+          session[:user_id] = ShopifyApp::SessionRepository.store_user_session(session_store, associated_user)
+          ShopifyApp::SessionRepository.update_scopes(shop_session, shop_scopes)
+        end
```
Comment (NabeelAhsen) [inline app/controllers/shopify_app/callback_controller.rb:104]: Is there an easy way to test this functionality?

**Review by NabeelAhsen — APPROVED** [review]
LGTM

We should reflect the following migration instruction in appropriate documentations:

> Run `rails g shopify_app:shop_model --skip`

**Comment by stevenmichaelthomas** [comment]
Can more context be added into this change to describe why it's needed, and what the impact could be without these changes? Thanks!

**Pull request closed without merging**